pkg/apiserver/internal/v1/offering.go

```diff
@@ -21,67 +21,133 @@ import (
+	offeringServer := &offeringServer{
+		client:        c,
+		dynamicClient: dynamicClient,
+		scheme:        scheme,
+	}
+	objGVK, err := apiutil.GVKForObject(&catalogv1alpha1.Offering{}, offeringServer.scheme)
+	if err != nil {
+		return nil, err
+	}
+	restMapping, err := restMapper.RESTMapping(objGVK.GroupKind(), objGVK.Version)
+	if err != nil {
+		return nil, err
 	}
+	offeringServer.gvr = restMapping.Resource
+	return offeringServer, nil
 }
 
 func (o offeringServer) List(ctx context.Context, req *v1.OfferingListRequest) (res *v1.OfferingList, err error) {
 	var listOptions []client.ListOption
 	listOptions, err = o.validateListRequest(req)
 	if err != nil {
-		return nil, status.Errorf(codes.InvalidArgument, err.Error())
+		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
 	offeringList := &catalogv1alpha1.OfferingList{}
 	if err := o.client.List(ctx, offeringList, listOptions...); err != nil {
-		return nil, status.Errorf(codes.Internal, fmt.Sprintf("listing offerings: %s", err.Error()))
+		return nil, status.Errorf(codes.Internal, "listing offerings: %s", err.Error())
 	}
 
 	res, err = o.convertOfferingList(offeringList)
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, fmt.Sprintf("converting OfferingList: %s", err.Error()))
+		return nil, status.Errorf(codes.Internal, "converting OfferingList: %s", err.Error())
 	}
 	return
 }
 
 func (o offeringServer) Get(ctx context.Context, req *v1.OfferingGetRequest) (res *v1.Offering, err error) {
 	if err = o.validateGetRequest(req); err != nil {
-		return nil, status.Errorf(codes.InvalidArgument, err.Error())
+		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
 	offering := &catalogv1alpha1.Offering{}
 	if err = o.client.Get(ctx, types.NamespacedName{
 		Name:      req.Name,
 		Namespace: req.Account,
 	}, offering); err != nil {
-		return nil, status.Errorf(codes.Internal, fmt.Sprintf("getting offering: %s", err.Error()))
+		return nil, status.Errorf(codes.Internal, "getting Offering: %s", err.Error())
 	}
 	res, err = o.convertOffering(offering)
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, fmt.Sprintf("converting Offering: %s", err.Error()))
+		return nil, status.Errorf(codes.Internal, "converting Offering: %s", err.Error())
 	}
 	return
 }
 
+func (o offeringServer) Watch(req *v1.OfferingWatchRequest, stream v1.OfferingService_WatchServer) error {
+	listOptions, err := o.validateWatchRequest(req)
+	if err != nil {
+		return status.Error(codes.InvalidArgument, err.Error())
+	}
+	watcher, err := o.dynamicClient.Resource(o.gvr).Namespace(req.Account).Watch(listOptions)
+	if err != nil {
+		return status.Errorf(codes.Internal, "watching offerings: %s", err.Error())
+	}
+	for {
+		select {
+		case <-stream.Context().Done():
+			return status.Error(codes.Internal, stream.Context().Err().Error())
```

Comment:
```suggestion
			return stream.Context().Err()
```

this would utilize the right error code, from the documentation:

```
	// Done returns a channel that's closed when work done on behalf of this
	// context should be canceled. Done may return nil if this context can
	// never be canceled. Successive calls to Done return the same value.
	// The close of the Done channel may happen asynchronously,
	// after the cancel function returns.
	//
	// WithCancel arranges for Done to be closed when cancel is called;
	// WithDeadline arranges for Done to be closed when the deadline
	// expires; WithTimeout arranges for Done to be closed when the timeout
	// elapses.
	//
	// Done is provided for use in select statements:
	//
	//  // Stream generates values with DoSomething and sends them to out
	//  // until DoSomething returns an error or ctx.Done is closed.
	//  func Stream(ctx context.Context, out chan<- Value) error {
	//  	for {
	//  		v, err := DoSomething(ctx)
	//  		if err != nil {
	//  			return err
	//  		}
	//  		select {
	//  		case <-ctx.Done():
	//  			return ctx.Err()
	//  		case out <- v:
	//  		}
	//  	}
	//  }
	//
	// See https://blog.golang.org/pipelines for more examples of how to use
	// a Done channel for cancellation.
	Done() <-chan struct{}
```

Thus if the user cancels the watch, the error code should be `codes.Cancelled = 1` which this approach satisfies.

Reply:
done